crates/kitsune_p2p/types/src/tx2/tx2_api.rs

```diff
@@ -0,0 +1,489 @@
+    fn eq(&self, oth: &Self) -> bool {
+        self.0.eq(&oth.0)
+    }
+}
+
+impl<C: Codec + 'static + Send + Unpin> Eq for Tx2EpHnd<C> {}
+
+impl<C: Codec + 'static + Send + Unpin> std::hash::Hash for Tx2EpHnd<C> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+impl<C: Codec + 'static + Send + Unpin> Tx2EpHnd<C> {
+    /// Capture a debugging internal state dump.
+    pub fn debug(&self) -> serde_json::Value {
+        self.0.debug()
+    }
+
+    /// Is this endpoint closed?
+    pub fn is_closed(&self) -> bool {
+        self.0.is_closed()
+    }
+
+    /// Close this endpoint.
+    pub fn close(
+        &self,
+        code: u32,
+        reason: &str,
+    ) -> impl std::future::Future<Output = ()> + 'static + Send {
+        self.0.close(code, reason)
+    }
+
+    /// Get the bound local address of this endpoint.
+    pub fn local_addr(&self) -> KitsuneResult<TxUrl> {
+        self.0.local_addr()
+    }
+
+    /// Establish a new connection.
+    pub fn connect<U: Into<TxUrl>>(
+        &self,
+        remote: U,
+        timeout: KitsuneTimeout,
+    ) -> impl std::future::Future<Output = KitsuneResult<Tx2ConHnd<C>>> + 'static + Send {
+        let rmap = self.1.clone();
+        self.0
+            .connect(remote.into(), timeout)
+            .map_ok(move |con| Tx2ConHnd::new(con, rmap))
+    }
+}
+
+/// Data associated with an IncomingConnection EpEvent
+#[derive(Debug)]
+pub struct Tx2EpIncomingConnection<C: Codec + 'static + Send + Unpin> {
+    /// the remote connection handle (could be closed)
+    pub con: Tx2ConHnd<C>,
+
+    /// the remote url from which this data originated
+    /// this is included incase the con is closed
+    pub url: TxUrl,
+}
+
+/// Respond to a Tx2EpIncomingRequest
+pub struct Tx2Respond<C: Codec + 'static + Send + Unpin> {
+    con: ConHnd,
+    msg_id: u64,
+    _p: std::marker::PhantomData<C>,
+}
+
+impl<C: Codec + 'static + Send + Unpin> std::fmt::Debug for Tx2Respond<C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("Tx2Respond").finish()
+    }
+}
+
+impl<C: Codec + 'static + Send + Unpin> Tx2Respond<C> {
+    fn new(con: ConHnd, msg_id: u64) -> Self {
+        Self {
+            con,
+            msg_id,
+            _p: std::marker::PhantomData,
+        }
+    }
+
+    /// Respond to a Tx2EpIncomingRequest
+    pub fn respond(
+        self,
+        data: C,
+        timeout: KitsuneTimeout,
+    ) -> impl std::future::Future<Output = KitsuneResult<()>> + 'static + Send {
+        let Tx2Respond { con, msg_id, .. } = self;
+        async move {
+            let mut buf = PoolBuf::new();
+            data.encode(&mut buf).map_err(KitsuneError::other)?;
+
+            // ignore errors, the remote con may have closed
+            // but that's their loss.
+            let _ = con.write(MsgId::new(msg_id).as_res(), buf, timeout).await;
+
```

Comment:
This seems ok but I wonder if it's worth a trace or debug level log?

Reply:
the fn here already returns a result type, so i just went ahead and bubbled the errors up here.